### DISCUSSION TITLE
Dynamic prompts - what am I doing wrong?

### DISCUSSION THREAD

**Discussion by tiko13** [vladmandic/sdnext#1521] [Q&A] [answered]
Hello guys

in A1111, when using prompt syntax of [A|B], the sampler alternates between using A and B each step, resulting in a kind of smooth blend of A and B. Here, it seems that the parser randomly chooses one of A/B and uses that as general prompt.

Please is there a way to enable the 'alternating' functionality?

**Comment by brknsoul** [comment, marked as answer]
~~First, Dynamic Prompts uses curly braces, not square brackets; `{like|this}`. Each generation will randomly select between `like` and `this`.~~

~~Second, If you want to generate all possible combinations, rather than just a random pick each time, tick the "Combinatorial generation" box under Dynamic Prompts.~~

Prompting with [A|B] will result in a combination of A and B, ran on alternating steps, useful, for example, if you want to combine two people or things; `[Angelina Jolie|Milla Jovovich]` to create a new person that has the features of both. ~~This combining feature doesn't require Dynamic Prompts, as it's an inbuilt function.~~

**Comment by tiko13** [comment]
Sorry for confusion, i didnt mean dynamic prompt extension, but rather the latter example [Angelina Jolie|Milla Jovovich] - for some reason this doesnt work in my case

**Reply by Flerndip** [reply]
I thought you had to use {} not [] to encapsulate non-wildcard variables i.e. {Angelina|Milla}.  Anyway I've been noticing the same thing.  Either|Or options are blending instead of using one and ignoring the others.  Example: {patent pumps|ballet shoes} and you get a lot of patent ballet shoes with stiletto heels in the results.

**Reply by brknsoul** [reply]
Please read the entire topic before posting. The OP isn't talking about the Dynamic Prompts extension.
Morphing uses square brackets, not curly braces.

**Reply by tiko13** [reply]
I have done some more testing today and indeed the alternate step rendering [X|Y] works fine, confirmed by checking preview every step. Its just that in case of colors, most of the time one seems to 'take over' and completely suppress the other one. I am not sure why this happens (it is not sampler specific anyway), so this approach is not very good for color 'mixing'

**Reply by brknsoul** [reply]
You can use the method `[red:blue:0.7]` to have `red` run in 70% of steps, then `blue` to run the other 30% of steps. Note that's a colon (`:`) between the two prompts, not a vertical line. 

This can be modified for single prompts;
`[red:0.7]` means run `red` until 70% of steps are complete, then stop running `red`
`[blue::0.7]` means don't run `blue` until 70% of steps are complete, then include `blue`

You can use a float for a percentage (eg `[red:0.7]` = `red` for 70% of steps)
or an integer for an exact number of steps (eg `[blue:15]` = `blue` for exactly 15 steps).

**Reply by tiko13** [reply]
Thanks, i knew about the double syntax [red:blue:15] or [red:blue:0.5] but I didnt know you can do it also with single token, that could be pretty neat

**Comment by brknsoul** [comment]
Seems to work for me;
`[Milla Jovovich|Kat Dennings]`
![image](https://github.com/vladmandic/automatic/assets/1969381/46c7f0fc-1db1-4422-b763-23144cee3cca)
 That's seems to be Jovovich's eyes and Dennings' lips. ;-)

Another test;
`a close shot photo of a 24 year old woman with [red|blue] hair`
The left one is at 25 steps, the right at 50 steps.
![image](https://github.com/vladmandic/automatic/assets/1969381/f96fa817-f957-475c-9158-63049b872e48)![image](https://github.com/vladmandic/automatic/assets/1969381/ef6cd0bf-9e0a-4009-8167-7aa4e5425c96)

**Reply by tiko13** [reply]
I have tried your prompt (DDIM, 30+30 steps):

a close shot photo of a 24 year old woman with red hair:
![image](https://github.com/vladmandic/automatic/assets/52453245/47914d88-9c39-48d0-ab4f-d10527175bc4)


a close shot photo of a 24 year old woman with [blonde|red] hair:
![image](https://github.com/vladmandic/automatic/assets/52453245/944ab9b3-131e-4ee1-935f-0f3d2ea8c2c5)

not sure what is going on, i didnt really touch the prompt settings

**Reply by brknsoul** [reply]
In Settings > Stable Diffusion, what "Prompt attention parser" are you using?

**Reply by tiko13** [reply]
I am using 'Full parser', tried A1111 parser just now but getting similar results

**Reply by brknsoul** [reply]
Does this happen with all samplers? (Mine were done using DPM++ 2M Karras)

**Reply by tiko13** [reply]
this is the result i get from A1111 for [blonde|red] hair
![image](https://github.com/vladmandic/automatic/assets/52453245/1f85fdd0-0fb6-4120-9a5e-f654de9fa5e4)

**Reply by brknsoul** [reply]
I'd keep using the Full Parser.

Another option is to use [x AND y] (the "AND" is case-specific.. needs to be ALL CAPS).
`A [cat AND dog]` results in this abomination;
![image](https://github.com/vladmandic/automatic/assets/1969381/d37660b7-5953-4b06-bd67-418ffe72a701)

**Comment by brknsoul** [comment]
Perhaps joining the SD.Next discord could get more people onto this?

**Reply by tiko13** [reply]
Thank you, i will ask there, I just fired up A1111 again (took ages lol) and the same prompt clearly works with the 'blending'
![image](https://github.com/vladmandic/automatic/assets/52453245/3d62bd7f-052b-41ce-a5e2-95c739f51d36)

**Reply by brknsoul** [reply]
I'd try two different colours, rather than two natural colours; like red and blue, or yellow and green.

**Reply by tiko13** [reply]
I have tried to install a fresh install of SD.Next and it works, there, so i will have to find out the cause in my extensions/settings it seems

**Reply by Aptronymist** [reply]
> Perhaps joining the SD.Next discord could get more people onto this?

*sigh* another thing I have on my list to write up.
The a1111 version is such a huge long document. 
I have considered writing this all up for SD.Next, this just makes it more imperative.
Gives me an idea though...
